akka-cluster/src/main/scala/akka/cluster/Gossip.scala

```diff
@@ -222,14 +292,60 @@ private[cluster] final case class Gossip(
     members.maxBy(m ⇒ if (m.upNumber == Int.MaxValue) 0 else m.upNumber)
   }
 
+  def removeAll(nodes: Iterable[UniqueAddress], removalTimestamp: Long): Gossip = {
+    nodes.foldLeft(this)((gossip, node) ⇒ gossip.remove(node, removalTimestamp))
+  }
+
+  def update(updatedMembers: immutable.SortedSet[Member]): Gossip = {
+    copy(members = updatedMembers union members)
+  }
+
+  /**
+   * Remove the given member from the set of members and mark it's removal with a tombstone to avoid having it
+   * reintroduced when merging with another gossip that has not seen the removal.
+   */
+  def remove(node: UniqueAddress, removalTimestamp: Long): Gossip = {
+    // removing REMOVED nodes from the `seen` table
+    val newSeen = overview.seen - node
+    // removing REMOVED nodes from the `reachability` table
+    val newReachability = overview.reachability.remove(node :: Nil)
+    val newOverview = overview.copy(seen = newSeen, reachability = newReachability)
+
+    // Clear the VectorClock when member is removed. The change made by the leader is stamped
+    // and will propagate as is if there are no other changes on other nodes.
+    // If other concurrent changes on other nodes (e.g. join) the pruning is also
+    // taken care of when receiving gossips.
+    val newVersion = version.prune(VectorClock.Node(ClusterCoreDaemon.vclockName(node)))
+    val newMembers = members.filterNot(_.uniqueAddress == node)
+    val newTombstones = tombstones + (node → removalTimestamp)
+    copy(version = newVersion, members = newMembers, overview = newOverview, tombstones = newTombstones)
+  }
+
+  def markAsDown(member: Member): Gossip = {
+    // replace member (changed status)
+    val newMembers = members - member + member.copy(status = Down)
+    // remove nodes marked as DOWN from the `seen` table
+    val newSeen = overview.seen - member.uniqueAddress
+
+    // update gossip overview
+    val newOverview = overview copy (seen = newSeen)
+    copy(members = newMembers, overview = newOverview) // update gossip
+  }
+
   def prune(removedNode: VectorClock.Node): Gossip = {
     val newVersion = version.prune(removedNode)
     if (newVersion eq version) this
     else copy(version = newVersion)
   }
 
+  def pruneTombstones(removeEarlierThan: Gossip.Timestamp): Gossip = {
+    val newTombstones = tombstones.filter { case (_, timestamp) ⇒ timestamp < removeEarlierThan }
```

Comment:
`filter`/`filterNot` you have once again beat me... 😠 Additionally a typo in the unit test made it pass